firebase-auth/src/commonMain/kotlin/dev/gitlive/firebase/auth/credentials.kt

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2020 GitLive Ltd.  Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package dev.gitlive.firebase.auth
+
+import dev.gitlive.firebase.Firebase
+
+expect open class AuthCredential {
+    val providerId: String
+}
+expect class PhoneAuthCredential : AuthCredential
+
+expect class OAuthCredential : AuthCredential
+
+expect object EmailAuthProvider {
+    fun credentialWithEmail(email: String, password: String): AuthCredential
+}
+
+expect object FacebookAuthProvider {
+    fun credentialWithAccessToken(accessToken: String): AuthCredential
+}
+
+expect object GithubAuthProvider {
+    fun credentialWithToken(token: String): AuthCredential
+}
+
+expect object GoogleAuthProvider {
+    fun credentialWithIDAndAccessToken(idToken: String, accessToken: String): AuthCredential
+}
+
+expect class OAuthProvider constructor(provider: String, auth: FirebaseAuth = Firebase.auth) {
+    companion object {
+        fun credentialsWithAccessToken(providerId: String, accessToken: String): AuthCredential
+        fun credentialsWithIDAndAccessToken(providerId: String, idToken: String, accessToken: String): AuthCredential
+        fun credentialsWithIDRawNonceAndAccessToken(providerId: String, idToken: String, rawNonce: String, accessToken: String): AuthCredential
+        fun credentialsWithIDAndRawNonce(providerId: String, idToken: String, rawNonce: String): AuthCredential
```

Comment:
https://github.com/GitLiveApp/firebase-kotlin-sdk#named-arguments *With... methods should have their With... suffix removed as named arguments in kotlin provide the distinction between overloaded function calls (like ObjectiveC/Swift). We can keep the *With.. names in @JvmName and @JsName annotations to prevent clashes on those platforms

Reply:
Since @JvmName and @JsName did not resolve clashes with relation to method signatures, I ssolved this by passing a Sealed class to the constructor instead.

Reply:
Yeah I got this wrong: "as named arguments in kotlin provide the distinction between overloaded function calls (like ObjectiveC/Swift)", not sure why I thought this - has this changed in 1.4 or it is still the case you can't have two methods with the same name and arg types?

Reply:
I think it the latter. Documentation only shows examples when dealing with generics. Anyway, its solved regardless

Reply:
tbh I think:
`OAuthProvider.credentialsWithIDAndRawNonce(...)`
is a more straightforward API than:
`OAuthProvider.credentials(IdAndAccessTokenAndRawNonce(...))`
so Im going to revert this change

Reply:
In the meantime I have queried the language design decision that prevents this.. https://discuss.kotlinlang.org/t/feature-request-support-method-overloads-on-parameter-names-like-swift/19577